python/pfs/utils/optimize_black_dots.py

```diff
@@ -0,0 +1,482 @@
+            x and y coordinate of the new position
+        """
+        dots = self.dots
+
+        xd_original = dots.loc[i-1]['x']
+        yd_original = dots.loc[i-1]['y']
+
+        xd_offset = xd_original - x_scale_rot
+        yd_offset = yd_original - y_scale_rot
+
+        r_distance_from_center = np.sqrt(xd_offset**2 + yd_offset**2)
+
+        xc_rotated, yc_rotated = self.rotate((xd_offset*(1+scale*r_distance_from_center/100),
+                                              yd_offset*(1+scale*r_distance_from_center/100)), (0, 0), rot)
+
+        xc_rotated = xc_rotated+x_scale_rot
+        yc_rotated = yc_rotated+y_scale_rot
+        # print(xc_rotated , yc_rotated)
+        xd_new = xc_rotated + xd_mod
+        yd_new = yc_rotated + yd_mod
+        return xd_new, yd_new
+
+    def quality_measure_single_spot(self, list_of_prepared_observations_and_predictions_single_fiber,
+                                    xd, yd):
+        """calculate a penalty given x and y dot positions
+
+        calls `check`
+
+        gets called by `total_penalty_for_single_dot`
+
+        Parameters
+        ----------
+        list_of_prepared_observations_and_predictions: `list`
+            list contaning observations for only one fiber
+        xd: `float`
+            x position of the dot
+        yd: `float`
+            y position of the dot
+
+        Returns
+        ----------
+        quality_measure: `float`
+            total penalty for this position of the dot
+        """
+        list_of_quality_measure_single_run = []
+        for i in range(len(list_of_prepared_observations_and_predictions_single_fiber)):
+            prepared_observations_and_predictions =\
+                list_of_prepared_observations_and_predictions_single_fiber[i]
+
+            actual_position_x_observed,\
+                actual_position_y_observed,\
+                predicted_position_x_not_observed,\
+                predicted_position_y_not_observed = prepared_observations_and_predictions
+
+            # the ones which were seen - add penalty if you should not have seen them
+            quality_measure_observed = []
+            for j in range(len(actual_position_x_observed)):
+                quality_measure_observed.append(self.check(actual_position_x_observed[j],
+                                                           actual_position_y_observed[j], xd, yd, 0.75))
+
+            # the ones which should not be seen - add penalty if you saw them
+            quality_measure_predicted = []
+            for j in range(len(predicted_position_x_not_observed)):
+                # if you were not able to predict position, no penalty
+                if np.isnan(predicted_position_x_not_observed[j]):
+                    val = 0
+                else:
+                    val = self.check(predicted_position_x_not_observed[j],
+                                     predicted_position_y_not_observed[j], xd, yd, 0.75)
+                    val = 1 - val
```

Comment:
They could be different, but it is not clear why would one weight them differently? Were you thinking that seen points should be weighted more strongly? 
I like that each point has a weight of 1, because in this case the final results tell you exactly how many points you got ``wrong'', i.e., it is easily understandable.